posthog/api/survey.py

```diff
@@ -751,6 +773,249 @@ def responses_count(self, request: request.Request, **kwargs):
+                "last_seen": None,
+            },
+            SurveyEventName.DISMISSED: {
+                "total_count": 0,
+                "unique_persons": 0,
+                "first_seen": None,
+                "last_seen": None,
+            },
+            SurveyEventName.SENT: {
+                "total_count": 0,
+                "unique_persons": 0,
+                "first_seen": None,
+                "last_seen": None,
+            },
+        }
+
+        # Update stats with actual results
+        for event_name, total_count, unique_persons, first_seen, last_seen in results:
+            if event_name in [e.value for e in SurveyEventName]:
+                stats[SurveyEventName(event_name)] = {
+                    "total_count": total_count,
+                    "unique_persons": unique_persons,
+                    "first_seen": first_seen.isoformat() + "Z" if first_seen else None,
+                    "last_seen": last_seen.isoformat() + "Z" if last_seen else None,
+                }
+        return stats
+
+    def _calculate_rates(self, stats: SurveyStats) -> SurveyRates:
+        """Calculate response and dismissal rates from stats.
+
+        Args:
+            stats: Dictionary containing event stats
+
+        Returns:
+            Dictionary containing calculated rates
+        """
+        rates: SurveyRates = {
+            "response_rate": 0.0,
+            "dismissal_rate": 0.0,
+        }
+
+        shown_count = stats.get(SurveyEventName.SHOWN, {}).get("total_count", 0)
+        if shown_count > 0:
+            sent_count = stats.get(SurveyEventName.SENT, {}).get("total_count", 0)
+            dismissed_count = stats.get(SurveyEventName.DISMISSED, {}).get("total_count", 0)
+            rates = {
+                "response_rate": round(sent_count / shown_count * 100, 2),
+                "dismissal_rate": round(dismissed_count / shown_count * 100, 2),
+            }
+        return rates
+
+    def _get_survey_stats(self, date_from: str | None, date_to: str | None, survey_id: str | None = None) -> dict:
+        """Get survey statistics from ClickHouse.
+
+        Args:
+            date_from: Optional ISO timestamp for start date with timezone info
+            date_to: Optional ISO timestamp for end date with timezone info
+            survey_id: Optional survey ID to filter for. If None, gets stats for all surveys.
+
+        Returns:
+            Dictionary containing survey statistics and rates
+        """
+        parsed_from, parsed_to = self._validate_and_parse_dates(date_from, date_to)
+
+        # Build cache key based on parameters
+        cache_key_parts = [f"survey_stats_{self.team_id}"]
+        if survey_id:
+            cache_key_parts.append(str(survey_id))
+        # Use parsed dates for cache key to ensure consistent caching regardless of input timezone
+        cache_key_parts.extend([str(parsed_from), str(parsed_to)])
+        cache_key = "_".join(cache_key_parts)
```

Comment:
cache_key is safe @ioannisj because for the team_id comes from the internals, survey_id is parsed to see if it exists, and the dates are validated too